Python should be configured/built with --enable-shared option.

Currently the Python installations are built with a command like:

```
RUN mkdir -p /usr/src/python \
    && curl -SL "https://www.python.org/ftp/python/$PYTHON_VERSION/Python-$PYTHON_VERSION.tar.xz" \
        | tar -xJC /usr/src/python --strip-components=1 \
    && cd /usr/src/python \
    && ./configure \
    && make -j$(nproc) \
    && make install \
    && cd / \
    && rm -rf /usr/src/python
```

When running `configure`, you should be supplying the `--enable-shared` option to ensure that shared libraries are built for Python. By not doing this you are preventing any application which wants to use Python as an embedded environment from working. This is because the lack of the shared library results in any embedded system failing at compile time with:

```
/usr/bin/ld: /usr/local/lib/libpython2.7.a(abstract.o): relocation R_X86_64_32S against `_Py_NotImplementedStruct' can not be used when making a shared object; recompile with -fPIC

/usr/local/lib/libpython2.7.a: error adding symbols: Bad value

collect2: error: ld returned 1 exit status

error: command 'gcc' failed with exit status 1
```

This basic mistake is something that Linux distributions themselves made for many years and it took a lot of complaining and education to get them to fix their Python installations. It would be nice to see you address this and do what all decent Linux distributions do now, and have done for a while, and install Python with shared libraries.

--- a/2.7/Dockerfile
+++ b/2.7/Dockerfile
@@ -1,9 +1,7 @@
-FROM buildpack-deps
-
-RUN apt-get update && apt-get install -y curl procps
+FROM buildpack-deps:jessie
 
 # remove several traces of debian python
-RUN apt-get purge -y python python-minimal python2.7-minimal
+RUN apt-get purge -y python.*
 
 # http://bugs.python.org/issue19846
 # > At the moment, setting "LANG=C" on a Linux system *fundamentally breaks Python 3*, and that's not OK.
@@ -11,18 +9,23 @@ ENV LANG C.UTF-8
 
 ENV PYTHON_VERSION 2.7.8
 
-RUN mkdir -p /usr/src/python \
+RUN set -x \
+	&& mkdir -p /usr/src/python \
 	&& curl -SL "https://www.python.org/ftp/python/$PYTHON_VERSION/Python-$PYTHON_VERSION.tar.xz" \
 		| tar -xJC /usr/src/python --strip-components=1 \
 	&& cd /usr/src/python \
-	&& ./configure \
+	&& ./configure --enable-shared \
 	&& make -j$(nproc) \
 	&& make install \
-	&& cd / \
+	&& ldconfig \
+	&& curl -SL 'https://bootstrap.pypa.io/get-pip.py' | python2 \
+	&& find /usr/local \
+		\( -type d -a -name test -o -name tests \) \
+		-o \( -type f -a -name '*.pyc' -o -name '*.pyo' \) \
+		-exec rm -rf '{}' + \
 	&& rm -rf /usr/src/python
 
-# install "pip" and "virtualenv", since the vast majority of users of this image will want it
-RUN curl -SL 'https://bootstrap.pypa.io/get-pip.py' | python2
+# install "virtualenv", since the vast majority of users of this image will want it
 RUN pip install virtualenv
 
 CMD ["python2"]

--- a/3.3/Dockerfile
+++ b/3.3/Dockerfile
@@ -1,9 +1,7 @@
-FROM buildpack-deps
-
-RUN apt-get update && apt-get install -y curl procps
+FROM buildpack-deps:jessie
 
 # remove several traces of debian python
-RUN apt-get purge -y python python-minimal python2.7-minimal
+RUN apt-get purge -y python.*
 
 # http://bugs.python.org/issue19846
 # > At the moment, setting "LANG=C" on a Linux system *fundamentally breaks Python 3*, and that's not OK.
@@ -11,24 +9,27 @@ ENV LANG C.UTF-8
 
 ENV PYTHON_VERSION 3.3.6
 
-RUN mkdir -p /usr/src/python \
+RUN set -x \
+	&& mkdir -p /usr/src/python \
 	&& curl -SL "https://www.python.org/ftp/python/$PYTHON_VERSION/Python-$PYTHON_VERSION.tar.xz" \
 		| tar -xJC /usr/src/python --strip-components=1 \
 	&& cd /usr/src/python \
-	&& ./configure \
+	&& ./configure --enable-shared \
 	&& make -j$(nproc) \
 	&& make install \
-	&& cd / \
+	&& ldconfig \
+	&& curl -SL 'https://bootstrap.pypa.io/get-pip.py' | python3 \
+	&& find /usr/local \
+		\( -type d -a -name test -o -name tests \) \
+		-o \( -type f -a -name '*.pyc' -o -name '*.pyo' \) \
+		-exec rm -rf '{}' + \
 	&& rm -rf /usr/src/python
 
 # make some useful symlinks that are expected to exist
 RUN cd /usr/local/bin \
-	&& ln -s easy_install-3.4 easy_install \
 	&& ln -s idle3 idle \
 	&& ln -s pydoc3 pydoc \
 	&& ln -s python3 python \
 	&& ln -s python-config3 python-config
-
-RUN curl -SL 'https://bootstrap.pypa.io/get-pip.py' | python3
 
 CMD ["python3"]

--- a/3.4/Dockerfile
+++ b/3.4/Dockerfile
@@ -1,9 +1,7 @@
-FROM buildpack-deps
-
-RUN apt-get update && apt-get install -y curl procps
+FROM buildpack-deps:jessie
 
 # remove several traces of debian python
-RUN apt-get purge -y python python-minimal python2.7-minimal
+RUN apt-get purge -y python.*
 
 # http://bugs.python.org/issue19846
 # > At the moment, setting "LANG=C" on a Linux system *fundamentally breaks Python 3*, and that's not OK.
@@ -11,14 +9,19 @@ ENV LANG C.UTF-8
 
 ENV PYTHON_VERSION 3.4.2
 
-RUN mkdir -p /usr/src/python \
+RUN set -x \
+	&& mkdir -p /usr/src/python \
 	&& curl -SL "https://www.python.org/ftp/python/$PYTHON_VERSION/Python-$PYTHON_VERSION.tar.xz" \
 		| tar -xJC /usr/src/python --strip-components=1 \
 	&& cd /usr/src/python \
-	&& ./configure \
+	&& ./configure --enable-shared \
 	&& make -j$(nproc) \
 	&& make install \
-	&& cd / \
+	&& ldconfig \
+	&& find /usr/local \
+		\( -type d -a -name test -o -name tests \) \
+		-o \( -type f -a -name '*.pyc' -o -name '*.pyo' \) \
+		-exec rm -rf '{}' + \
 	&& rm -rf /usr/src/python
 
 # make some useful symlinks that are expected to exist

